command/cmd_copy.go

```diff
@@ -0,0 +1,638 @@
+	// suffix.
+	if dst.IsRemote() {
+		if strings.HasSuffix(dst.Path, "/") {
+			dst = dst.Join(objname)
+		}
+		return dst, nil
+	}
+
+	client, err := storage.NewClient(dst)
+	if err != nil {
+		return nil, err
+	}
+
+	// For local->local copy operations, we can safely stat <dst> to check if
+	// it is a file or a directory.
+	obj, err := client.Stat(ctx, dst)
+	if err != nil && err != storage.ErrGivenObjectNotFound {
+		return nil, err
+	}
+
+	// Absolute <src> path is given. Use given <dst> and local copy operation
+	// will create missing directories if <dst> has one.
+	if !originalSrc.HasGlob() {
+		return dst, nil
+	}
+
+	// For local->local copy operations, if <src> has glob, <dst> is expected
+	// to be a directory. As always, local copy operation will create missing
+	// directories if <dst> has one.
+	if obj != nil && !obj.Type.IsDir() {
+		return nil, fmt.Errorf("destination argument is expected to be a directory")
+	}
+
+	return dst.Join(objname), nil
+}
+
+// prepareDownloadDestination will return a new destination URL for
+// remote->local and remote->remote copy operations.
+func prepareDownloadDestination(
+	ctx context.Context,
+	originalSrc *objurl.ObjectURL,
+	src *objurl.ObjectURL,
+	dst *objurl.ObjectURL,
+	parents bool,
+) (*objurl.ObjectURL, error) {
+	objname := src.Base()
+	if parents {
+		objname = src.Relative()
+	}
+
+	if originalSrc.HasGlob() {
+		if err := os.MkdirAll(dst.Absolute(), os.ModePerm); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := storage.NewClient(dst)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := client.Stat(ctx, dst)
+	if err != nil && err != storage.ErrGivenObjectNotFound {
+		return nil, err
+	}
+
+	if parents {
+		if obj != nil && !obj.Type.IsDir() {
+			return nil, fmt.Errorf("destination argument is expected to be a directory")
+		}
+		dst = dst.Join(objname)
+		if err := os.MkdirAll(dst.Dir(), os.ModePerm); err != nil {
+			return nil, err
+		}
+	}
+
+	if err == storage.ErrGivenObjectNotFound {
+		if err := os.MkdirAll(dst.Dir(), os.ModePerm); err != nil {
+			return nil, err
+		}
+		if strings.HasSuffix(dst.Absolute(), "/") {
+			dst = dst.Join(objname)
+		}
+	} else {
+		if obj.Type.IsDir() {
+			dst = obj.URL.Join(objname)
+		}
+	}
+
+	return dst, nil
+}
+
+// prepareUploadDestination will return a new destination URL for local->remote
+// operations.
+func prepareUploadDestination(
+	src *objurl.ObjectURL,
+	dst *objurl.ObjectURL,
+	parents bool,
+) *objurl.ObjectURL {
+	objname := src.Base()
+	if parents {
+		objname = src.Relative()
+	}
+	return dst.Join(objname)
+}
+
+// expandSource returns the full list of objects from the given src argument.
+// If src is an expandable URL, such as directory, prefix or a glob, all
+// objects are returned by walking the source.
+func expandSource(
+	ctx context.Context,
+	src *objurl.ObjectURL,
+	isRecursive bool,
+) (<-chan *storage.Object, error) {
+	// TODO(ig): this function could be in the storage layer.
```

Comment:
+1

Reply:
not going to happen in this PR. looks easy but i think it's not.